### PR TITLE
Multi line begin

### PR DESCRIPTION
**Pull request by luisggpina**
This PR adds support for multi-line begin.

I changed the API a bit:  I deleted the `begin` function.  The begin sequence is now part of `match`.  When there is a begin sequence, all entries in the `start_state` array point to the begin entry-point.  When the begin sequence gets matched, it rewrites the `start_state` array with the correct entry-points.

For example, the following rule:

```
begin => open(_, _, _),
         mmap(_, _, _, _, _, _),
         close(_)
```

Results in:

```
match(remote=getpid, local=*all but open*) => EXECUTE
match(remote=getpid, local=open)           => EXECUTE
match(remote=getpid, local=mmap)           => EXECUTE
match(remote=getpid, local=close)          => EXECUTE (begin finished here)
match(remote=getpid, local=getpid)         => SELF_DONE
```

Note how the begin sequence is now part of the `match` function.  The caller (Varan) doesn't even know there is a begin sequence.  This makes more sense and is a better separation of concerns than having a `begin` function like we originally suggested.

### CODE REVIEW

File: src/CodeGen.hs [hunk 1]
```diff
@@ -213,9 +207,15 @@ initArray lst f func def array
                             Nothing -> def)
            in (idx + 1, d <+> (indent 2 $ text (array ++ "[" ++ (show idx) ++ "] = " ++ t ++ ";\n")))
 
+initFpsBody :: [Mapping] -> Doc
+initFpsBody mappings
+  = initArray mappings (\(mapping,_) -> mapping) "&init_" "&init_default_match" "start_state"
```
Comment (daniel-grumberg) [inline src/CodeGen.hs]: Instead of `(\(mapping, _) -> mapping)` use `const`. I think I wrote this, but can you change it?

File: src/CodeGen.hs [hunk 2]
```diff
@@ -389,6 +402,35 @@ makeFpValidation (Self args mnem code) info
...
 
+makeFpBegin :: RhsElem -> (String, Int, Int) -> Bool -> Doc
+makeFpBegin (Rhs (sys@(Name sysname), _, _, _)) info last
+  = text ("static enum %action% " ++ fname
+          ++ "(%state_t% *state, struct %syscall% *local) {") <$$>
+    indent 2 (text $ "if ((local->syscall != " ++ nr_s ++") || (!" 
+                      ++ fname ++"_val(local))) {") <$$>
+    (if is_first then
+        indent 4 (text $ "state->state_func = &" ++ fname ++ ";") <$$>
+        indent 4 (text "return EXECUTE;")
```
Comment (daniel-grumberg) [inline src/CodeGen.hs]: Why do we return `EXECUTE` here, as we failed to match local with the start of the begin sequence? Or is the logic here, that as long as we do not match the start of the begin sequence anything goes as it is assumed to be part of the setup code of the follower?

Comment (luisggpina) [inline src/CodeGen.hs]: Exactly, begin accepts everything until it matches the sequence.  Then, it reverts back to the rules in the DSL file.

File: src/CodeGen.hs (hunk 1 above)
Comment (luisggpina) [inline src/CodeGen.hs]: Done that, and rebased the changes so this code is outdated.

**Comment by luisggpina** [comment]
Fixes #4

**Pull request merged**